core/message_pool.go

```diff
@@ -12,6 +12,19 @@ import (
 	"github.com/filecoin-project/go-filecoin/types"
 )
 
+// MessageTimeOut is the number of tipsets we should receive before timing out messages
+const MessageTimeOut = 6
+
+type timedmessage struct {
+	message *types.SignedMessage
+	addedAt uint64
+}
+
+// BlockTimer defines a interface to a struct that can give the current block height.
+type BlockTimer interface {
+	BlockHeight() (uint64, error)
```

Comment:
FYI In my WIP message queue I use the term "stamp" to refer to the time-like mark associated with each message. It's opaque to the queue/pool and should make no difference if the stamps and age limit were converted to seconds.

So this could then become `Stamper` with `CurrentStamp()` method, no reference to "blocks" or height etc. The wrapper struct could be `StampedMessage. The interpretation as block height is done by the user/constructor that hooks things up.

Reply:
The fact that this is a BlockHeight is central to invalidation logic. It can't be abstracted away. See comment on the role of `MessagePool`.